File: ch14/FindStringsByRegex.java
/* ###
 * IP: GHIDRA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// Counts the number of defined strings that match a regex in the current
// selection, or current program if no selection is made, and displays the
// results on the console
//
//@author Ghidrabook, KN 
//@category Ghidrabook.CH14
//@keybinding 
//@menupath 
//@toolbar 

import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.*;
import ghidra.program.util.ProgramSelection;

import java.io.*;

public class FindStringsByRegex extends GhidraScript {

	@Override
	public void run() throws Exception {
		String regex =
			askString("Please enter the regex",
				      "Please enter the regex you're looking to match:");
		
		Listing listing = currentProgram.getListing();
		
		DataIterator dataIt;
		if (currentSelection != null) {
			dataIt = listing.getDefinedData(currentSelection, true);
		} else {
			dataIt = listing.getDefinedData(true);
		}

		Data data;
		String type;
		int counter = 0;
		while (dataIt.hasNext() && !monitor.isCancelled()) {
			data = dataIt.next();
			type = data.getDataType().getName().toLowerCase();
			if (type.contains("unicode") || type.contains("string")) {
				String s = data.getDefaultValueRepresentation();
				if (s.matches(regex)) {
					counter++;
					println(s);
				}
			}
		}
		println(counter + " matching strings were found");
	}
}

File: ch14/FindStringsByRegex.py
# Counts the number of defined strings that match a regex in the current
# selection, or current program if no selection is made, and displays the
# results on the console
#@author Ghidrabook
#@category Ghidrabook.CH14
#@keybinding 
#@menupath 
#@toolbar
#
# Copyright (c) 2019 Kara Nance (knance@securityworks.com)
# 
# Permission is hereby granted, free of charge, to any person obtaining a copy of 
# this software and associated documentation files (the "Software"), to deal in 
# the Software without restriction, including without limitation the rights to 
# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of 
# the Software, and to permit persons to whom the Software is furnished to do so, 
# subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in all 
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS 
# FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR 
# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER 
# IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN 
# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

import re

regex = askString(
    "RegEx", "Please enter the regex you're looking to match:")

listing = currentProgram.getListing()

if currentSelection != None:
    dataIt = listing.getDefinedData(currentSelection, True)
else:
    dataIt = listing.getDefinedData(True)

counter = 0;
while dataIt.hasNext() and not monitor.isCancelled():
    data = dataIt.next()
    data_type = data.getDataType().getName().lower()
    if 'unicode' in data_type or 'string' in data_type:
        s = data.getDefaultValueRepresentation()
        if re.match(regex, s) is not None:
            counter+=1
            print(s)

print('%d matching strings were found' % (counter, ))